### ISSUE TITLE
Threading

### ISSUE THREAD

**Issue by astryzia** [astryzia/bash-portscanner#20]
Runs a bit slow without the backgrounding method for multithreading. Scanning for a large number of ports can quickly exhaust system resources if all are backgrounded. Look into the best middle ground for threading.

**Comment by muschem** [comment]
Ugh....hadn't checked this yet. It looks pretty bad - around an order of magnitude slower than nmap in simple test cases I just tried. Probably my fault for breaking performance. You had a clean method before.

**Comment by muschem** [comment]
Ok, some data points from my test lab. Trying to eliminate variables and compare apples to apples. These numbers are for 7 targets in a /24 network segment, using nmap's top 1,000 ports:

Test case #1 = nmap:
real	0m3.224s
user	0m0.095s
sys	0m0.846s

Test case #2 = script v0.0.1 (https://github.com/serializingme/bash-portscanner/blob/fc1fb14e201c977a2e68b7773789b2878b8629e1/portscan.sh):
real	0m15.099s
user	0m17.713s
sys	0m8.121s

Test case #3 = current version:
real	0m26.686s
user	0m18.003s
sys	0m3.889s

So, comparing to nmap wasn't exactly fair in my earlier comment, as we weren't that performant to begin with. We have added some significant overhead, but it wasn't quite as drastic an increase as I thought at first. Still definitely worth exploring for performance enhancements.

**Comment by muschem** [comment]
So, after a bit more testing, backgrounding the portscan process does save significant time in many cases (at least on my test systems). It also breaks the method I was using for storing and reporting the open ports, though. Hopefully, a fix won't be too problematic for that - I'll do a little testing later to see if I can work something out.

**Comment by muschem** [comment]
Ok, I have a working model for parallelizing the workload. It works great for a single host - working out bugs in multi-host reporting. I'll need to re-think how timing delays should be restructured for parallel workloads, because this approach partially breaks the very basic method I used for sleep delays in a completely serialized model. All of it should be solvable with a bit more work.

**Comment by muschem** [comment]
Partial solution offered in PR #22 
This doesn't completely fix the problem, as work is still needed for banner grabbing and parallelizing other functions in general, but for straight up port scanning, it is much better. I'll work on testing timing delays, because I think this added perf may break the naive approach I was taking with that.

**Comment by astryzia** [comment]
Fantastic so far. I had no idea you could do this sort of thing with bash. TIL! I've been testing timing delays, and the perf function does seem to break them.

**Comment by muschem** [comment]
Yeah, the more parallel the workload, the more broken the current timing method becomes, I would think. In order to increase performance, the parallelexec function attempts to always keep a certain number of pids running (exact amount determined by a limiting factor applied to output of ulimit -u). Each of the commands for port scanning is prepended with a sleep call (duration determined by timing setting), which works fine if they are all done in serial fashion, back to back. But, when a large group is started all at roughly the same time, those sleep calls also expire at roughly the same time, meaning the target gets hammered by a bunch of nearly-simultaneous scans, breaking the intent of the timing delay. 

One thought I had was to mutate the sleep delay with a random function including some upper and lower bounds. So, each scan process would sleep a different amount of time up to a certain defined maximum. This might help prevent, a large block of scan processes from all initiating at nearly the same time. But, I need more coffee and a bit of number crunching to figure out how to set those bounds... the scaling should probably vary based on the width of the parallelized workload, but that gets a bit complex.

**Comment by astryzia** [comment]
Where are we at with this now? I'll dig through code but wanted to ask here so we can work through issues with threading and tie it off :)

**Comment by muschem** [comment]
Pretty much the same spot. I had intended to work on this more, but got a bit sidetracked with other projects.

**Comment by astryzia** [comment]
No worries! I'm getting back to a point where I can actively help with this again so I'll take a look at it
